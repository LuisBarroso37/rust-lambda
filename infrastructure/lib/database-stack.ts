import type { StackProps } from "aws-cdk-lib";
import { CfnOutput, RemovalPolicy, Stack } from "aws-cdk-lib";
import { AttributeType, BillingMode, Table } from "aws-cdk-lib/aws-dynamodb";
import type { Construct } from "constructs";

export class DatabaseStack extends Stack {
	constructor(scope: Construct, id: string, properties?: StackProps) {
		super(scope, id, properties);

		const table = new Table(this, "Table", {
			partitionKey: {
				name: "#p",
				type: AttributeType.STRING,
			},
			sortKey: {
				name: "#s",
				type: AttributeType.STRING,
			},
			billingMode: BillingMode.PAY_PER_REQUEST,
			pointInTimeRecovery: true,
			deletionProtection: true,
		});

		table.applyRemovalPolicy(RemovalPolicy.RETAIN_ON_UPDATE_OR_DELETE);

		new CfnOutput(this, "TableName", {
			description: "The name of the DynamoDB table",
			value: table.tableName,
			exportName: "Table-TableName",
		});

		new CfnOutput(this, "TableArn", {
			description: "The ARN of the DynamoDB table",
			value: table.tableArn,
			exportName: "Table-TableArn",
		});
	}
}
