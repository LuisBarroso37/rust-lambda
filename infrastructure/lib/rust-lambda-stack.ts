import * as path from "path";
import { CfnOutput, Fn, RemovalPolicy, Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { RustFunction } from "cargo-lambda-cdk";
import { Architecture, FunctionUrlAuthType } from "aws-cdk-lib/aws-lambda";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import { PolicyStatement } from "aws-cdk-lib/aws-iam";

export class RustLambdaStack extends Stack {
	constructor(scope: Construct, id: string, props?: StackProps) {
		super(scope, id, props);

		const tableArn = Fn.importValue("Table-TableArn");
		const tableName = Fn.importValue("Table-TableName");

		const environment = {
			TABLE_NAME: tableName,
		};

		const dynamodbReadPolicy = new PolicyStatement({
			actions: ["dynamodb:Query", "dynamodb:GetItem", "dynamodb:Scan"],
			resources: [tableArn, `${tableArn}/index/*`],
		});
		const dynamodbReadWritePolicy = new PolicyStatement({
			actions: [
				"dynamodb:Query",
				"dynamodb:GetItem",
				"dynamodb:Scan",
				"dynamodb:PutItem",
				"dynamodb:UpdateItem",
			],
			resources: [tableArn, `${tableArn}/index/*`],
		});

		const getAllFunction = new RustFunction(this, "GetAllFunction", {
			manifestPath: path.join(__dirname, "..", "..", "rust_lambda"),
			binaryName: "get_all",
			runtime: "provided.al2023",
			architecture: Architecture.ARM_64,
			environment,
		});

		getAllFunction.addToRolePolicy(dynamodbReadPolicy);

		new LogGroup(this, "LogGroup1", {
			logGroupName: `/aws/lambda/${getAllFunction.functionName}`,
			retention: RetentionDays.TEN_YEARS,
			removalPolicy: RemovalPolicy.RETAIN,
		});

		const insertFunction = new RustFunction(this, "InsertFunction", {
			manifestPath: path.join(__dirname, "..", "..", "rust_lambda"),
			binaryName: "insert",
			runtime: "provided.al2023",
			architecture: Architecture.ARM_64,
			environment,
		});

		insertFunction.addToRolePolicy(dynamodbReadWritePolicy);

		new LogGroup(this, "LogGroup2", {
			logGroupName: `/aws/lambda/${insertFunction.functionName}`,
			retention: RetentionDays.TEN_YEARS,
			removalPolicy: RemovalPolicy.RETAIN,
		});

		const getAllFunctionUrl = getAllFunction.addFunctionUrl({
			authType: FunctionUrlAuthType.NONE,
		});

		const insertFunctionUrl = insertFunction.addFunctionUrl({
			authType: FunctionUrlAuthType.NONE,
		});

		new CfnOutput(this, "GetAllFunctionUrl", {
			value: getAllFunctionUrl.url,
		});

		new CfnOutput(this, "InsertFunctionUrl", {
			value: insertFunctionUrl.url,
		});
	}
}
