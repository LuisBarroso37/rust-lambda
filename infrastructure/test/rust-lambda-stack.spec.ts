import { expect, test } from "@jest/globals";
import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";

import { RustLambdaStack } from "../lib/rust-lambda-stack";
import { cleanupUnnecessaryChecksForSnapshot } from "./utils";

test("Access Stack Snapshot Test", () => {
	const app = new App();
	const stack = new RustLambdaStack(app, "RustLambdaStack");
	const template = Template.fromStack(stack);

	expect(cleanupUnnecessaryChecksForSnapshot(template)).toMatchSnapshot();
});
