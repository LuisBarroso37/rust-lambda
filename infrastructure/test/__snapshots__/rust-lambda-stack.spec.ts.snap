// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Access Stack Snapshot Test 1`] = `
{
  "Outputs": {
    "GetAllFunctionUrl": {
      "Value": {
        "Fn::GetAtt": [
          "GetAllFunctionFunctionUrl73FAD014",
          "FunctionUrl",
        ],
      },
    },
    "InsertFunctionUrl": {
      "Value": {
        "Fn::GetAtt": [
          "InsertFunctionFunctionUrlABAD530D",
          "FunctionUrl",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "GetAllFunction07A130BE": {
      "DependsOn": [
        "GetAllFunctionServiceRoleDefaultPolicyA595CA2C",
        "GetAllFunctionServiceRole0BEC2B19",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "PLACEHOLDER",
        },
        "Handler": "bootstrap",
        "Role": {
          "Fn::GetAtt": [
            "GetAllFunctionServiceRole0BEC2B19",
            "Arn",
          ],
        },
        "Runtime": "provided.al2023",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GetAllFunctionFunctionUrl73FAD014": {
      "Properties": {
        "AuthType": "NONE",
        "TargetFunctionArn": {
          "Fn::GetAtt": [
            "GetAllFunction07A130BE",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Url",
    },
    "GetAllFunctionServiceRole0BEC2B19": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GetAllFunctionServiceRoleDefaultPolicyA595CA2C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:Query",
                "dynamodb:GetItem",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::ImportValue": "Table-TableArn",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::ImportValue": "Table-TableArn",
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetAllFunctionServiceRoleDefaultPolicyA595CA2C",
        "Roles": [
          {
            "Ref": "GetAllFunctionServiceRole0BEC2B19",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetAllFunctioninvokefunctionurl08074081": {
      "Properties": {
        "Action": "lambda:InvokeFunctionUrl",
        "FunctionName": {
          "Fn::GetAtt": [
            "GetAllFunction07A130BE",
            "Arn",
          ],
        },
        "FunctionUrlAuthType": "NONE",
        "Principal": "*",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "InsertFunction44D5A36F": {
      "DependsOn": [
        "InsertFunctionServiceRoleDefaultPolicy1911E434",
        "InsertFunctionServiceRole92CA57BB",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "PLACEHOLDER",
        },
        "Handler": "bootstrap",
        "Role": {
          "Fn::GetAtt": [
            "InsertFunctionServiceRole92CA57BB",
            "Arn",
          ],
        },
        "Runtime": "provided.al2023",
      },
      "Type": "AWS::Lambda::Function",
    },
    "InsertFunctionFunctionUrlABAD530D": {
      "Properties": {
        "AuthType": "NONE",
        "TargetFunctionArn": {
          "Fn::GetAtt": [
            "InsertFunction44D5A36F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Url",
    },
    "InsertFunctionServiceRole92CA57BB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "InsertFunctionServiceRoleDefaultPolicy1911E434": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::ImportValue": "Table-TableArn",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::ImportValue": "Table-TableArn",
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "InsertFunctionServiceRoleDefaultPolicy1911E434",
        "Roles": [
          {
            "Ref": "InsertFunctionServiceRole92CA57BB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "InsertFunctioninvokefunctionurlE487BB5D": {
      "Properties": {
        "Action": "lambda:InvokeFunctionUrl",
        "FunctionName": {
          "Fn::GetAtt": [
            "InsertFunction44D5A36F",
            "Arn",
          ],
        },
        "FunctionUrlAuthType": "NONE",
        "Principal": "*",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "LogGroup106AAD846": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "GetAllFunction07A130BE",
              },
            ],
          ],
        },
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "LogGroup2477F707C": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "InsertFunction44D5A36F",
              },
            ],
          ],
        },
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
